Source/Private/ASTextNode2.mm

```diff
@@ -0,0 +1,1122 @@
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    textLayoutCache = [[NSCache alloc] init];
+  });
+  
+  ASTextCacheValue *cacheValue = ({
+    std::lock_guard<std::mutex> lock(layoutCacheLock);
+    cacheValue = [textLayoutCache objectForKey:text];
+    if (cacheValue == nil) {
+      cacheValue = [[ASTextCacheValue alloc] init];
+      [textLayoutCache setObject:cacheValue forKey:text];
+    }
+    cacheValue;
+  });
+  
+  CGRect containerBounds = (CGRect){ .size = container.size };
+  {
+    std::lock_guard<std::mutex> lock(cacheValue->_m);
+    for (auto &t : cacheValue->_layouts) {
+      CGSize constrainedSize = std::get<0>(t);
+      ASTextLayout *layout = std::get<1>(t);
+      
+      CGSize layoutSize = layout.textBoundingSize;
+      // 1. CoreText can return frames that are narrower than the constrained width, for obvious reasons.
+      // 2. CoreText can return frames that are slightly wider than the constrained width, for some reason.
+      //    We have to trust that somehow it's OK to try and draw within our size constraint, despite the return value.
+      // 3. Thus, those two values (constrained width & returned width) form a range, where
+      //    intermediate values in that range will be snapped. Thus, we can use a given layout as long as our
+      //    width is in that range, between the min and max of those two values.
+      CGRect minRect = CGRectMake(0, 0, MIN(layoutSize.width, constrainedSize.width), MIN(layoutSize.height, constrainedSize.height));
+      if (!CGRectContainsRect(containerBounds, minRect)) {
+        continue;
+      }
+      CGRect maxRect = CGRectMake(0, 0, MAX(layoutSize.width, constrainedSize.width), MAX(layoutSize.height, constrainedSize.height));
+      if (!CGRectContainsRect(maxRect, containerBounds)) {
+        continue;
+      }
+      
+      // Now check container params.
+      ASTextContainer *otherContainer = layout.container;
+      if (!UIEdgeInsetsEqualToEdgeInsets(container.insets, otherContainer.insets)) {
+        continue;
+      }
+      if (!ASObjectIsEqual(container.exclusionPaths, otherContainer.exclusionPaths)) {
+        continue;
+      }
+      if (container.maximumNumberOfRows != otherContainer.maximumNumberOfRows) {
+        continue;
+      }
+      if (container.truncationType != otherContainer.truncationType) {
+        continue;
+      }
+      if (!ASObjectIsEqual(container.truncationToken, otherContainer.truncationToken)) {
+        continue;
+      }
+      // TODO: When we get a cache hit, move this entry to the front (LRU).
+      return layout;
+    }
+  }
+  
+  // Cache Miss.
+  
+  // Compute the text layout.
+  ASTextLayout *layout = [ASTextLayout layoutWithContainer:container text:text];
+  
+  // Store the result in the cache.
+  {
+    std::lock_guard<std::mutex> lock(cacheValue->_m);
+    cacheValue->_layouts.push_front(std::make_tuple(container.size, layout));
+    if (cacheValue->_layouts.size() > 3) {
```

Comment:
> Is 3 a magic number?
Yes.
> I wonder if there's value in having control of this. My guess is specific strings have layouts that would be worth caching and most have 0

No probably not. This cache is primarily intended to serve between layout passes and displays of the same text node. Any benefit that we get across the app is negligible (but great!).